turbustat/statistics/pca/width_estimate.py

```diff
@@ -0,0 +1,387 @@
+
+        if method == 'fit':
+            output, cov = fit_2D_gaussian(xmat, ymat, z)
+            y_scales[idx] = output.y_stddev_0.value
+            x_scales[idx] = output.x_stddev_0.value
+
+            errs = np.sqrt(np.abs(cov.diagonal()))
+            # Order in the cov matrix is given by the order of parameters in
+            # model.param_names. But amplitude and the means are fixed, so in
+            # this case they are the first 2.
+            y_scale_errors[idx] = errs[1]
+            x_scale_errors[idx] = errs[0]
+
+            if diagnosticplots and idx < 9:
+                import matplotlib.pyplot as plt
+                ax = plt.subplot(3, 3, idx + 1)
+                ax.imshow(z, cmap='afmhot')
+                ax.contour(output(xmat, ymat),
+                           levels=np.array([0.25, 0.5, 0.75, 1.0]) * z.max(),
+                           colors=['c'] * 3)
+                # ax.show()
+
+        elif method == 'interpolate':
+            warn("Error estimation not implemented for interpolation!")
+            rvec = rmat.ravel()
+            zvec = z.ravel()
+            zvec /= zvec.max()
+            sortidx = np.argsort(zvec)
+            rvec = rvec[sortidx]
+            zvec = zvec[sortidx]
+            dz = len(zvec) / 100.
+            spl = LSQUnivariateSpline(zvec, rvec, zvec[dz:-dz:dz])
+
+            x_scales[idx] = spl(np.exp(-1)) / np.sqrt(2)
+            y_scales[idx] = spl(np.exp(-1)) / np.sqrt(2)
+
+            # Need to implement some error estimation
+            x_scale_errors[idx] = 0.0
+            y_scale_errors[idx] = 0.0
+
+        elif method == 'xinterpolate':
+            warn("Error estimation not implemented for interpolation!")
+            output, cov = fit_2D_gaussian(xmat, ymat, z)
+            aspect = output.y_stddev_0.value[0] / output.x_stddev_0.value[0]
+            theta = output.theta_0.value[0]
+            rmat = ((xmat * np.cos(theta) + ymat * np.sin(theta))**2 +
+                    (-xmat * np.sin(theta) + ymat * np.cos(theta))**2 *
+                    aspect**2)**0.5
+            rvec = rmat.ravel()
+            zvec = z.ravel()
+            zvec /= zvec.max()
+            sortidx = np.argsort(zvec)
+            rvec = rvec[sortidx]
+            zvec = zvec[sortidx]
+            dz = len(zvec) / 100.
+            spl = LSQUnivariateSpline(zvec, rvec, zvec[dz:-dz:dz])
+
+            x_scales[idx] = spl(np.exp(-1)) / np.sqrt(2)
+            y_scales[idx] = spl(np.exp(-1)) / np.sqrt(2)
+
+            # Need to implement some error estimation
+            x_scale_errors[idx] = 0.0
+            y_scale_errors[idx] = 0.0
+
+        elif method == 'contour':
+            znorm = z
+            znorm /= znorm.max()
+
+            # Import contour tools and paths that don't require making a
+            # plotting window
+            # http://www.dalkescientific.com/writings/diary/archive/2005/04/23/matplotlib_without_gui.html
+            import matplotlib._cntr as cntr
```

Comment:
No kidding! I'm glad this worked [as described](http://stackoverflow.com/questions/5666056/matplotlib-extracting-data-from-contour-lines) because I really didn't want to go hunting through the bowels of matplotlib.